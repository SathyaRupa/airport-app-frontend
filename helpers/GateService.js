import axios from 'axios';
import url from './Url';

export default {
  fetchAll: async page => {
    try {
      const response = await axios.get(`${url}/gates?page=${page}`);
      return response.data;
    } catch (error) {
      console.log('Error fetching data: ', error);
    }
  },
};
