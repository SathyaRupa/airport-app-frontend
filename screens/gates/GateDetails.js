import React, {useEffect, useState} from 'react';
import {View, Text, StyleSheet} from 'react-native';
import GateService from '../../helpers/GateService';
import {Card} from 'react-native-paper';
import LinearGradient from 'react-native-linear-gradient';

function GateDetails({route}) {
  const {id} = route.params;
  const [details, setDetails] = useState({});

  useEffect(() => {
    GateService.show(id)
      .then(response => {
        setDetails(response);
      })
      .catch(error => {
        console.error('Error fetching Gate details:', error);
      });
  }, [id]);

  return (
    <View style={styles.container}>
      <Card mode="elevated" style={styles.card}>
        <LinearGradient
          colors={['#4D869C', '#7AB2B2']}
          start={{x: 1, y: 0}}
          end={{x: 1, y: 1}}
          style={styles.gradient}>
          <Card.Content>
            <View style={styles.content}>
              <Text style={styles.info}>
                Gate Number : {details.gate_number}
              </Text>
              <Text style={styles.info}>
                Floor Number : {details.floor_number}
              </Text>
            </View>
          </Card.Content>
        </LinearGradient>
      </Card>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  card: {
    width: 300,
    height: 200,
  },
  gradient: {
    borderRadius: 12,
    overflow: 'hidden',
    width: '100%',
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
  content: {
    justifyContent: 'center',
  },
  info: {
    fontSize: 18,
    marginBottom: 5,
    color: 'white',
  },
  default: {
    fontSize: 12,
    fontWeight: 'bold',
    justifyContent: 'center',
  },
});

export default GateDetails;
