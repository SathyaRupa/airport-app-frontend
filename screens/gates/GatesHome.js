import ItemCard from '../../components/ItemCard';
import * as React from 'react';
import {useEffect, useState} from 'react';
import GateService from '../../helpers/GateService';
import {FlatList, ActivityIndicator} from 'react-native';
import {Avatar} from 'react-native-paper';
import {useIsFocused} from '@react-navigation/native';

function GatesHome({navigation}) {
  const [allDataLoaded, setAllDataLoaded] = useState(false);
  const [fetchedPages, setFetchedPages] = useState([]);
  const [gates, setGates] = useState([]);
  const [loading, setLoading] = useState(true);
  const [page, setPage] = useState(0);
  const isFocused = useIsFocused();

  useEffect(() => {
    async function fetchData() {
      if (!allDataLoaded && isFocused && !fetchedPages.includes(page)) {
        const response = await GateService.fetchAll(page);
        setFetchedPages(prevPages => [...prevPages, page]);
        if (response.length === 0) {
          setAllDataLoaded(true);
        } else {
          setGates(prevGates => [...prevGates, ...response]);
        }
      }
      setLoading(false);
    }
    fetchData();
  }, [page, allDataLoaded, isFocused]);

  const loadMoreGates = () => {
    if (!allDataLoaded) {
      setPage(prevPage => prevPage + 1);
    }
  };

  useEffect(() => {
    if (isFocused) {
      setGates([]);
      setPage(0);
      setAllDataLoaded(false);
      setLoading(true);
      setFetchedPages([]);
    }
  }, [isFocused]);

  const renderFooter = () => {
    return loading ? <ActivityIndicator size="large" color="#4D869C" /> : null;
  };

  const handlePress = id => {
    navigation.push('Gate Details', {id});
  };

  const icon = (
    <Avatar.Image
      style={{backgroundColor: 'transparent'}}
      size={50}
      source={require('../../assets/icons/gates.png')}
    />
  );

  return (
    <FlatList
      data={gates}
      renderItem={itemData => (
        <ItemCard
          id={itemData.item.id}
          value={'Gate ' + itemData.item.gate_number}
          icon={icon}
          testId={`item-card-${itemData.index}`}
          onPress={() => handlePress(itemData.item.id)}
        />
      )}
      onEndReached={loadMoreGates}
      onEndReachedThreshold={0.5}
      ListFooterComponent={renderFooter}
    />
  );
}

export default GatesHome;
