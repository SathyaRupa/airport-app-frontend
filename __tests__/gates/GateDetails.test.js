import React from 'react';
import {fireEvent, render, waitFor, debug} from '@testing-library/react-native';
import GateService from '../../helpers/GateService';
import {useIsFocused} from '@react-navigation/native';
import GatesHome from '../../screens/Gates/GatesHome';
import GateDetails from '../../screens/gates/GateDetails';

jest.mock('react-native-linear-gradient', () => 'LinearGradient');
jest.mock('@react-navigation/native', () => ({
  useIsFocused: jest.fn(),
}));
jest.mock('../../helpers/GateService', () => ({
  fetchAll: jest.fn(),
  fetchGateInfo: jest.fn(),
}));

jest.mock('react-native-toast-message', () => 'Toast');
jest.mock('../../components/ToastMessage', () => ({
  SuccessToast: jest.fn(),
  ErrorToast: jest.fn(),
}));

describe('Gates home - get all Gates', () => {
  let mockNavigation;
  beforeEach(() => {
    mockNavigation = {
      push: jest.fn(),
    };
    jest.clearAllMocks();
    jest.useFakeTimers();
  });

  const mockRoute = {
    params: {
      id: 1,
    },
  };

  it('snapshot test', () => {
    GateService.fetchGateInfo.mockResolvedValueOnce([
      {floor_number: '10', gate_number: '17'},
    ]);
    const gateDetails = render(<GateDetails route={mockRoute} />);
    expect(gateDetails).toMatchSnapshot();
  });

  it('should render an item card and navigate to Gate Details page when item card is pressed', async () => {
    useIsFocused.mockReturnValue(true);
    GateService.fetchAll.mockResolvedValueOnce([
      {id: '1', floor_number: '10', gate_number: '17'},
      {id: '2', floor_number: '6', gate_number: '10'},
    ]);

    const {getAllByTestId} = render(<GatesHome navigation={mockNavigation} />);

    await waitFor(() => {
      const cardButton = getAllByTestId('card');
      expect(cardButton[0]).toBeTruthy();
      expect(mockNavigation.push).not.toHaveBeenCalled();
      fireEvent.press(cardButton[0]);
    });
    expect(mockNavigation.push).toHaveBeenCalledWith('Gate Details', {
      id: '1',
    });
  });
});
