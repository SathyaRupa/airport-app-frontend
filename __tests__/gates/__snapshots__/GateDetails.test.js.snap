// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gates home - get all Gates snapshot test 1`] = `
<Text>
  No details available.
</Text>
`;
